### ISSUE TITLE
pull googlecentric files (.gdoc, .gsheet, ...) like the windows drive client does

### ISSUE THREAD

**Issue by syp1975** [odeke-em/drive#904]
Hi, I would ask a minor enhancement.
On #404 says is not possible.
I think it means that the contents of the document is not downloadable and has to be exported.
What I would like is to download the files with a content like this:
```
{"url": "https://docs.google.com/open?id=1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8", "doc_id": "1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8", "email": "owner@example.com", "resource_id": "document:1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8"}
```
When the file is double-clicked, on Windows desktop clients with Chrome installed at least, it opens the document on google docs.

The modification would consist on treating the .gxxx files found on google drive as *regular* text files, without loosing the possibility to export the file.

Right now I use this script to convert .desktop files to .gxxx files:
`find . -name *.desktop -exec ~/desktop2url.sh -r '{}' \;`
desktop2url.sh:
```
#!/bin/bash
#case-insensitive
shopt -s nocaseglob
REMOVE=0
CLEAN=0
while test $# -gt 0; do
  case "$1" in
    -r|--remove)
      REMOVE=1
      shift
      ;;
    -c|--clean)
      CLEAN=1
      shift
      ;;
    *)
      break
      ;;
  esac
done

while test $# -gt 0; do
  FILE=${1}
  URL=$(grep -Pio '(?<=url=).+' "$FILE")
  DIRNAME=$(dirname "$FILE")
  FILENAME=`echo -n $(basename "$FILE") | cut -d'.' -f1`
  if [[ $URL =~ ^http[s]?://docs\.google\.com/.*spreadsheets/.* ]]; then
    OUTPUT="$DIRNAME/$FILENAME.gsheet"
    echo $OUTPUT
    echo -ne "{\"url\":\"$URL\"}" > "$OUTPUT"
  elif [[ $URL =~ ^http[s]?://docs\.google\.com/.*document/.* ]]; then
    OUTPUT="$DIRNAME/$FILENAME.gdoc"
    echo $OUTPUT
    echo -ne "{\"url\":\"$URL\"}" > "$OUTPUT"
  elif [[ $URL =~ ^http[s]?://docs\.google\.com/.*presentation/.* ]]; then
    OUTPUT="$DIRNAME/$FILENAME.gslides"
    echo $OUTPUT
    echo -ne "{\"url\":\"$URL\"}" > "$OUTPUT"
  elif [[ $URL =~ ^http[s]?://docs\.google\.com/.*drawings/.* ]]; then
    OUTPUT="$DIRNAME/$FILENAME.gdraw"
    echo $OUTPUT
    echo -ne "{\"url\":\"$URL\"}" > "$OUTPUT"
  elif [[ ! -z $URL ]]; then
    OUTPUT="$DIRNAME/$FILENAME.url"
    echo $OUTPUT
    echo -ne "[InternetShortcut]\nURL=$URL" > "$OUTPUT"
  fi
  if [[ ! -z $URL ]]; then
    [[ $REMOVE -gt 0 ]] && rm -f "$FILE"
    [[ $CLEAN -gt 0 ]] && rm -f "$DIRNAME/$FILENAME"
  fi
  shift
done
```

**Comment by odeke-em** [comment]
Hello there @syp1975, thank you for filing the issue and welcome to drive!

So currently the clickable files that you are talking about work only for Linux as mentioned at  https://github.com/odeke-em/drive#desktop-files. 

This looks possible and I can see you've got a format for *.gdoc files, then let's do it.

I don't think I have a Windows machine around but I'll take a look around my machine stash.

Also OSX users, if anyone knows how to make clickable files, let's get this rolling.

Thank you!

**Comment by shaggytwodope** [comment]
@odeke-em I'm not a mac user. Nor a great golang dev. But I've googled around some and found the reverse tool...but it may be a start. https://github.com/tischda/webloc I haven't found anything really detailed about .webloc files on mac. But I think this might be a start.

From my understanding the linux side generates a (web) link to the file on google drive. So effectively this would be what I believe to be .webloc files for mac. I hate anything thats not linux seesh... lol.

**Comment by odeke-em** [comment]
Thank you @shaggytwodope for starting the research #realMVP 🍻

So I have never heard of webloc, I wonder if webloc is a standard format?

**Comment by shaggytwodope** [comment]
@odeke-em so digging around more I've found this [blog post](http://kadin.sdf-us.org/weblog/technology/web/mac-webloc-files.html). And this [page](https://fileinfo.com/extension/webloc) will prolly sum it up far better then me.

Now this is just me throwing out what appears to be a step, but I'm not sure if this is at all the ideal way to do this on mac. Being such a Linux fanboy and all... I know next to nothing short of it's unix roots.

**Comment by odeke-em** [comment]
Thanks @shaggytwodope, I'll take another look at it.

So a friend just hinted at using .lnk but that didn't work so I instead looked at .url files e.g
```shell
$ cat sample.url
[InternetShortcut]
URL=https://github.com/odeke-em/drive
```
which then opened the URL in the default web browser for me when I clicked it, on a Windows 7 machine.

@syp1975 in regards to your request
```shell
{"url": "https://docs.google.com/open?id=1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8", "doc_id": "1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8", "email": "owner@example.com", "resource_id": "document:1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8"}
```
unfortunately we don't save the user's, we just only note down the respective OAuth2.0 granted tokens.

**Comment by syp1975** [comment]
@odeke-em, yes .url will work on Windows machines, but they are generic not googlecentric like .gdoc, .gsheet, .gslide, ... so the purpose of the feature request will not be fulfilled , that is, to mimic the windows drive client.
If you look at my script, you will see 2 things:

- it uses .url files as a fallback for unknown google document formats.
- .gxxx files are generated just with just the url property and they work as expected:
```
{"url": "https://docs.google.com/open?id=1M7_317tNGkSeaWLLXXXuPqyNQJAoe7vXGUhmH8bNov8"}
```

What about a `--url-files` option, and a `--gdoc-files` option?

@shaggytwodope, I'm the new IT manager at my job, and the first decision was to substitute a Windows Server with a Debian Linux Server, XD. Still, desktops and laptops run on Windows, and users expect the new server to work as before, but better.

**Comment by odeke-em** [comment]
Thank you for the update @syp1975, yes in deed I looked through your script. I tried out making *.gdoc files but on my machine, they required me to register a custom program to open them. Since the Windows client already does that, then like you mention, let's fall in line.

`
What about a --url-files option, and a --gdoc-files option?
`
Could you please explain what those options would do?
We currently have in pull
```shell
-desktop-links
    	allows docs + sheets to be pulled as .desktop files or URL linked files (default true)
```
which IMO would work for the purpose of exporting to the output format for the desktop files depending on the platform. 

So aside from the question that I have about your request for two different options ie --url-files and --gdoc-files, the information here looks actionable that we can get started with implementing it.
If you don't mind, I'll ping you on the next PR for this. It is 3:02AM here and am about to sleep so perhaps in the afternoon. However, if you'd like to send a PR, that would be very welcome, the code is in https://github.com/odeke-em/drive/blob/master/src/misc.go#L355-L380
<img width="986" alt="screen shot 2017-04-29 at 3 06 05 am" src="https://cloud.githubusercontent.com/assets/4898263/25554365/d3000278-2c88-11e7-868c-a60a20610dd5.png">


We have a few options:
a) Use Go's auto compilation selection per file i.e
make three separate files: misc.go, misc_linux.go, misc_windows.go and then set the respective build tags ie
```
+build linux
```
```
+build windows
```
then modify the `File.serializeAsDesktopEntry` method for the respective platforms.


b) Simpler option would be to modify that method to return different output and extension by platform as we've noted, different format for the .gdoc files for windows while a different one for Linux as .desktop.

Either way, IMO we are gucci to go!

**Comment by syp1975** [comment]
I think it would be nicer give more runtime options and to not limit this feature to the build platform, because for example, drive could run on a linux headless server, but stored files could be accessed by a linux, windows or macos desktop.
I explain:
```
--desktop-links
          allows docs + sheets to be pulled as .desktop files or URL linked files (default true)
--url-links
          allows docs + sheets to be pulled as .url files and be opened on desktop browsers (default false)
--gdoc-links
          allows docs + sheets to be pulled as .gxxx files and be opened on chrome (default false)
--webloc-links
          allows docs + sheets to be pulled as .webloc files for macos desktops (default false)
```
The syntax is up to you, maybe is easier add `url,gdoc,gsheet,gslide,webloc` to the list of supported `--export-formats`.

just FYI: the reason I run drive on the server is because of its pull and export capabilities, to do a full backup of the google drive and docs.

**Comment by odeke-em** [comment]
`
I think it would be nicer give more runtime options and to not limit this feature to the build platform, because for example, drive could run on a linux headless server, but stored files could be accessed by a linux, windows or macos desktop
`
That would be nice but I don't even know if many people even use the *.desktop files, if many people request for that feature, then perhaps we'll add it, but for now I think by platform exports suffice -- please feel free to file an issue after we've solved this one.
